### PR TITLE
fix: reject newlines in redirect locations

### PR DESCRIPTION
**Pull request by FrankChen021**
## What changed

- Validate the fully constructed redirect target before placing it in the `Location` response header.
- Reject redirect targets containing literal carriage returns or line feeds with HTTP 400.
- Preserve ordinary redirects and percent-encoded `%0D`/`%0A` URL data.
- Add focused tests for normal, encoded, CR, and LF cases.

## Why

`RedirectFilter` copied a URL derived from request data directly into the `Location` header. `java.net.URL` permits literal CR/LF characters, which could allow HTTP response splitting in servlet containers that do not reject them independently.

## Impact

Legitimate redirect targets retain their exact URL representation. Requests that would produce a raw newline in the response header are rejected before the response status or header is set.

## Validation

- `mvn -ntp test -pl server -am -Dtest=org.apache.druid.server.http.RedirectFilterTest -Dsurefire.failIfNoSpecifiedTests=false -Pskip-static-checks -Dweb.console.skip=true -T1C`
- `mvn -ntp test -pl server -Dtest=org.apache.druid.server.http.RedirectFilterTest -Dweb.console.skip=true -Pskip-static-checks`
- `mvn -ntp checkstyle:check -pl server -Dweb.console.skip=true`

### CODE REVIEW

File: server/src/test/java/org/apache/druid/server/http/RedirectFilterTest.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  public void testRedirectPreservesEncodedNewlines() throws Exception
+  {
+    final String location = "https://leader.example/path%0D%0Avalue?query=%0a";
+    Mockito.when(redirectInfo.getRedirectURL(QUERY_STRING, REQUEST_URI)).thenReturn(URI.create(location).toURL());
+
+    redirectFilter.doFilter(request, response, filterChain);
+
+    Mockito.verify(response).setStatus(HttpServletResponse.SC_TEMPORARY_REDIRECT);
+    Mockito.verify(response).setHeader("Location", location);
+  }
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply server/src/test/java/org/apache/druid/server/http/RedirectFilterTest.java:87]: Addressed in commit 6e192e4173: the encoded-newline redirect test now verifies that the filter chain is not invoked. RedirectFilterTest passes 4/4; server checkstyle and strict test compilation also pass.

[the remark this answers is not in the document]
Comment (FrankChen021) [reply server/src/test/java/org/apache/druid/server/http/RedirectFilterTest.java:87]: Addressed in 6e192e4173. The encoded-newline redirect test now verifies that the filter chain is not invoked.

File: server/src/main/java/org/apache/druid/server/http/RedirectFilter.java [hunk 2]
```diff
@@ -72,17 +72,23 @@ public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
...
       if (url == null) {
         // We apparently have nothing to redirect to, so let's do a Service Unavailable
         response.sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE);
         return;
       }
 
+      final String location = url.toString();
+      final String sanitizedLocation = location.replace('\r', ' ').replace('\n', ' ');
+      if (!location.equals(sanitizedLocation)) {
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST);
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply server/src/main/java/org/apache/druid/server/http/RedirectFilter.java:84]: Addressed in 6e192e4173. Redirect validation now checks CR and LF directly and sets Location from the original validated string.

File: server/src/test/java/org/apache/druid/server/http/RedirectFilterTest.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  private void assertInvalidRedirect(String location) throws Exception
+  {
+    Mockito.when(redirectInfo.getRedirectURL(QUERY_STRING, REQUEST_URI)).thenReturn(urlWithExternalForm(location));
+
+    redirectFilter.doFilter(request, response, filterChain);
+
+    Mockito.verify(response).sendError(HttpServletResponse.SC_BAD_REQUEST);
+    Mockito.verify(response, Mockito.never()).setStatus(Mockito.anyInt());
+    Mockito.verify(response, Mockito.never()).setHeader(Mockito.anyString(), Mockito.anyString());
+  }
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply server/src/test/java/org/apache/druid/server/http/RedirectFilterTest.java:111]: Addressed in commit 6e192e4173: invalid redirect targets now verify that the filter chain is not invoked after the 400 response. RedirectFilterTest passes 4/4; server checkstyle and strict test compilation also pass.

[the remark this answers is not in the document]
Comment (FrankChen021) [reply server/src/test/java/org/apache/druid/server/http/RedirectFilterTest.java:111]: Addressed in 6e192e4173. The invalid-redirect helper now verifies that the filter chain is not invoked after the 400 response.

File: server/src/main/java/org/apache/druid/server/http/RedirectFilter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (FrankChen021) [reply server/src/main/java/org/apache/druid/server/http/RedirectFilter.java:84]: Addressed in commit 6e192e4173: CR and LF are checked directly and the original Location value is used on the success path, avoiding the extra sanitized string allocation. RedirectFilterTest passes 4/4; server checkstyle and strict test compilation also pass.

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the code for correctness, edge cases, concurrency, and integration risks; no issues found.

Reviewed 2 of 2 changed files.

---

This is an automated review by Codex GPT-5.6-Sol

**Review by GWphua — APPROVED** [review]

**Review by GWphua — APPROVED** [review]

File: server/src/test/java/org/apache/druid/server/http/RedirectFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            throw new UnsupportedOperationException();
+          }
+
+          @Override
+          protected String toExternalForm(final URL url)
+          {
+            return externalForm;
+          }
+        }
+    );
```
[the remark this answers is not in the document]
Comment (FrankChen021) [reply server/src/test/java/org/apache/druid/server/http/RedirectFilterTest.java:131]: No code change is needed here. The current branch targets Java 25 (`java.version` and `maven.compiler.release` are both 25), and its CI matrix runs JDK 25. `URL.of(URI, URLStreamHandler)` is available for that target, and all 28 checks on this head pass. Resolving as not applicable to the supported Java version.

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the code for correctness, edge cases, concurrency, and integration risks; no issues found.

Reviewed 2 of 2 changed files. The previous baseline was unavailable locally; the full current diff was reviewed.



---

This is an automated review by Codex GPT-5.6-Luna(max)
